File: assignment-01/cpp/file1.cpp
#include<iostream>
#include <string>
using namespace std;

class matrix{
	
	private:
		int** p;
		int rows;
		int columns;
	public:
		matrix();
		~matrix();
		void setMatrix();
		void showMatrix();
		bool checkingEchelonForm();
		bool checkingReduceEchelonForm();

};

matrix::matrix(){
	cout<<"Enter Rows = ";
	cin>>rows;
	cout<<"Enter columns = ";
	cin>>columns;
	p = new int*[rows];
	for (int i = 0; i < rows; i++){
		p[i] = new int[columns];
		for (int j = 0; j < columns; j++)
		{
			p[i][j] = 0;
		}
	}
}

matrix::~matrix(){
	for (int i = 0; i < rows; i++){
		delete[] p[i];
	}
	delete[] p;
}

void matrix::showMatrix(){
	for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            cout << p[i][j] << " ";
        }
        cout<<endl;
    }
}

void matrix::setMatrix(){
	for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            cout << "Enter value at row = " << i << " column = "<<j<<" : ";
            cin>>p[i][j];
        }
    }
}

bool matrix::checkingEchelonForm()
{
	if (rows == 0 && columns == 0 )
	{
		// cout<<"Matrix is empty"<<endl;
		return false;
	}

	else if(rows == 1)
	{
		return true;
		// cout<<"Matrix is in Echelon Form"<<endl;
	}
	else{
		int lowestNumberOfZeros = -1;
		for (int i = 0; i < rows; i++)
		{
			int count = 0;
			for (int j = 0; j < columns; j++)
			{
				if (p[i][j] == 0)
				{
					count += 1;
				}
				else
				{
					break;
				}
			}
			if (count < lowestNumberOfZeros)
			{
				// cout<< "Matrix is not in Echelon Form"<<endl;
				return false;
			}
			else if (count == lowestNumberOfZeros)
			{
				if (count == columns)
				{
					continue;
				}
				else
				{
					// cout<<"Matrix is not in Echelon Form"<<endl;
					return false;
				}
			}
			else
			{
				lowestNumberOfZeros = count;
				continue;
			}

		}
		// cout<<"Matrix is in Echelon Form"<<endl;
		return true;
			
	}
}

bool matrix::checkingReduceEchelonForm()
{
	if (rows == 0 && columns == 0 )
	{
		// cout<<"Matrix is empty"<<endl;
		return false;
	}

	else if(rows == 1)
	{
		for (int i = 0; i < columns; ++i)
		{
			if (p[0][i] == 0)
			{
				continue;
			}
			else if (p[0][i] == 1)
			{
				return true;
			}
			else
			{
				return false;
			}
		}
		return true;
		// cout<<"Matrix is in  Reduce Echelon Form"<<endl;
	}
	else{
		int lowestNumberOfZeros = -1;
		for (int i = 0; i < rows; i++)
		{
			int count = 0;
			for (int j = 0; j < columns; j++)
			{
				if (p[i][j] == 0)
				{
					count += 1;
				}
				else if (p[i][j] == 1)
				{
					for (int k = 0; k < rows; k++)
					{
						if (p[k][j] == p[i][j])
						{
							continue;
						}
						else
						{ 
							if (p[k][j] == 0)
							{
								continue;
							}
							else
							{
								// cout<<"1"<<endl;
								return false;
							}
						}
					}
					break;
				}
				else
				{	
					
					// cout<<"2"<<endl;
					return false;
				}
			}
			if (count < lowestNumberOfZeros)
			{
				// cout<< "Matrix is not in Reduce Echelon Form"<<endl;
				return false;
			}
			else if (count == lowestNumberOfZeros)
			{
				if (count == columns)
				{
					continue;
				}
				else
				{
					// cout<<"Matrix is not in Reduce Echelon Form"<<endl;
					return false;
				}
			}
			else
			{
				lowestNumberOfZeros = count;
				continue;
			}

		}
		// cout<<"Matrix is in Reduce Echelon Form"<<endl;
		return true;
			
	}
}

int main(){

	matrix m;
	m.showMatrix();
	m.setMatrix();
	cout<<m.checkingReduceEchelonForm()<<endl;
	m.showMatrix(); 

	return 0;
}

File: assignment-01/python/file1.py
class matrix(object):
    """docstring for matrix"""
    def __init__(self, row, col):
        self.mat = []
        self.rows = row
        self.columns = col
        for i in range(0,self.rows):
            self.mat.append([])
            for j in range(0, self.columns):
                self.mat[i].append(0)

    def setDeepCopy(self,valMat):
        self.mat = []
        for i in range(0,self.rows):
            self.mat.append([])
            for j in range(0, self.columns):
                self.mat[i].append(valMat[i][j])


    def setShalowCopy(self,valMat):
    	self.mat = valMat




    def __str__(self):
        returnStr = ""
        for i in range(0,self.rows):
            for j in range(0,self.columns):
                returnStr += str(self.mat[i][j])
                returnStr += " "
            if i+1 != self.rows:
                returnStr += "\n"
        return returnStr 

    

    def countZeroInRow(self,arr):
    	count = 0
    	for i in range(0, self.columns):
    		if arr[i] != 0:
    			return count
    		else:
    			count += 1
    	return count


    def checkEchelonForm(self):
        if self.rows == self.columns == 0:
            return False
        elif self.rows == 1:
            return True
        else:
            lowestNumberOfZeros = -1
            for i in range(0, self.rows):
                numberOfZeros = self.countZeroInRow(self.mat[i])
                if lowestNumberOfZeros < numberOfZeros:
                    lowestNumberOfZeros = numberOfZeros
                elif lowestNumberOfZeros == numberOfZeros:
                    if numberOfZeros == self.columns:
                        continue
                    else:
                        return False
                else:
                    return False
            return True



    def checkReduceEchelonForm(self):
        if self.rows == self.columns == 0:
            return False
        elif self.rows == 1:
            return True
        else:
            lowestNumberOfZeros = -1
            for i in range(0, self.rows):
                numberOfZeros = 0
                for j in range(0, self.columns):
                    if self.mat[i][j] == 0:
                        numberOfZeros += 1
                    elif self.mat[i][j] == 1:
                        for k in range(0, self.rows):
                            if id(self.mat[k][j]) == id(self.mat[i][j]) or self.mat[k][j] == 0:
                                continue
                            else:
                                return False
                        break
                    else:
                        return False
                if lowestNumberOfZeros < numberOfZeros:
                    lowestNumberOfZeros = numberOfZeros
                elif lowestNumberOfZeros == numberOfZeros:
                    if numberOfZeros == self.columns:
                        continue
                    else:
                        return False
                else:
                    return False
            return True


    
m = matrix(3,3)
t = [[1,0,0],[0,1,0],[0,0,1]]
m.setDeepCopy(t)
# print(m.checkEchelonForm())
print(m.checkReduceEchelonForm())
print(m)